rfcs/20200710-reduction/README.md

```diff
@@ -0,0 +1,220 @@
+  ``` python
+  def l2_normalize(x, eps_mode, eps):
+      l2_norm = pow( eps_mode( sum([pow(x_i, 2) for x_i in x]), eps), 0.5)
+      return [x_i / l2_norm for x_i in x]
+  ```
+  where eps_mode = {add, max}
+* TensorFlow:
+  ``` python
+  def l2_normalize(x, eps):
+      l2_norm = pow( max( sum([pow(x_i, 2) for x_i in x]), eps), 0.5)
+      return [x_i / l2_norm for x_i in x]
+  ```
+* PyTorch:
+  ``` python
+  def l2_normalize(x, eps):
+      l2_norm = max( pow( sum([pow(x_i, 2) for x_i in x]), 0.5), eps)
+      return [x_i / l2_norm for x_i in x]
+  ```
+
+# Proposal
+
+## Common part: Dst memory descriptor.
+Dst memory descriptor is required to indicate reduction dimensions and dst data
+type. As a result dst memory descriptor also contains information about dst
+memory layout. Ideally reduction primitive should preserve memory layout as most
+oneDNN primitives to minimize amount of reorders between multiple layouts.
+However mixing of reduction operation and oneDNN blocked layouts leads to the
+following situations:
+* Let's assume user has memory in nChw16c layout and wants to reduce it over all
+ dimensions. Dst memory descriptor will have 4 dimensions similar to src, but
+ will contain ones in all dimensions. The main issue here is dst layout. Since
+ reduction should preserve layout dst will have nChw16c layout which is
+ illogical, because 16c guarantees that size c divides by 16 and oneDNN
+ internals rely on this. To address this issue dst can be padded to have 16
+ elements.
+* Let's assume user has memory in nChw16c layout and wants to reduce it over
+ dimensions h and w. As before dst memory descriptor will have 4 dimensions
+ similar to src, but will contain {n, c, 1, 1} dimensions. In this case layout
+ can be either nChw16c or nchw or nhwc and really depends on what user wants to
+ do with result.
+
+The following behavior is proposed to address these issues:
+* If any is provided for dst layout primitive preserves layout;
+* Otherwise primitives uses provided layout.
+
+## Option 1: reduction and norms as algorithms of single primitive (Recommended)
+
+### Pros
+* Minimizes amount of post operations required to implement framework operations,
+ hence implementation is simpler.
+
+### Cons
+* New algorithms on framework side will require changes on oneDNN side.
+
+### API
+``` cpp
+// dnnl_types.h
+typedef enum {
+    ...
+    dnnl_eltwise_pow,
+    ...
+    dnnl_reduction_max,
+    dnnl_reduction_min,
+    dnnl_reduction_norm_lp_max,
+    dnnl_reduction_norm_lp_add,
+    dnnl_reduction_norm_lp_power_p_max,
+    dnnl_reduction_norm_lp_power_p_add,
+    dnnl_reduction_sum,
+    dnnl_reduction_mul,
+    dnnl_reduction_mean,
+    ...
+} dnnl_alg_kind_t;
+
+// dst_desc is required to mark reduction dimensions and dst data type.
+typedef struct {
+    dnnl_primitive_kind_t primitive_kind;
+    // The kind of reduction algorithm.
+    dnnl_alg_kind_t reduction_alg_kind;
+    // Algorithm specific parameters.
+    float p, eps;
```

Comment:
nit: put in the end to align with bnorm, lnrom, lrn, eltwise, etc...

Reply:
Fixed